packages/ckeditor5-engine/src/model/selection.ts

```diff
@@ -667,11 +667,25 @@ export default class Selection extends /* #__PURE__ */ EmitterMixin( TypeCheckab
 				yield startBlock as any;
 			}
 
-			for ( const value of range.getWalker() ) {
+			const treewalker = range.getWalker();
+
+			for ( const value of treewalker ) {
 				const block = value.item;
 
 				if ( value.type == 'elementEnd' && isUnvisitedTopBlock( block as any, visited, range ) ) {
 					yield block as Element;
+				} else if ( block.is( 'model:element' ) && block.root.document!.model.schema.isBlock( block ) ) {
+					if ( value.type == 'elementEnd' ) {
+						treewalker.jumpTo( value.nextPosition );
+					} else {
+						const position = treewalker.position.clone();
+
+						position.offset = block.maxOffset;
+
+						if ( range.containsPosition( position ) ) {
```

Comment:
And if the position is not in range, then what?

I think it should be safe to `break;` here. Even if we find something inside, it will not be recorded because this will be the top block. Even more -- it shouldn't be recorded (at least according to the method's description), so if it is, it looks like a bug, right?

After the loop we check the final position anyway. It would be good to test it with some more interesting cases like:

```
<block>..[..</block>
<block>
	..
	<inlineElement>..]..</inlineElement>
	..
</block>
```

It should work fine, though, from the quick glance at `getParentBlock`.

Note, that we currently don't have inline elements that can include selection, text or other elements.